[3851] Path Navigator Issues with get Children API

aachavarria opened the pull request:
### Ticket reference or full description of what's in the PR

aachavarria committed:
ItemMenu.tsx: Fixing issue when there are no options for the item;
PathNavigator.tsx: adding debounce, and adding pagination support;
ToolsPanelEmbeddedAppViewButton.tsx: Fixing issue when mixins.tools;

aachavarria committed:
Implementing StateMap

aachavarria committed:
Fixing issue with getChildrenByPath() and null levelDescriptor

rart commented on ui/app/src/components/ItemMenu/ItemMenu.tsx:
```diff
@@ -159,7 +159,7 @@ function ItemMenuUI(props: ItemMenuUIProps) {
     isValidCutPastePath(item.path, clipboard.sourcePath);
   const options = generateSingleItemOptions(item, { hasClipboard, ...permissions });
-  const noOptions = options.length === 1 && options[0].length === 0;
+  const noOptions = options.length === 0 || (options.length === 1 && options[0].length === 0);
```
What's the meaning of this condition? Add comment describing

rart commented on ui/app/src/components/Navigation/PathNavigator/PathNavigator.tsx:
```diff
@@ -141,15 +152,23 @@ export default function PathNavigator(props: PathNavigatorProps) {
+  useEffect(() => {
+    onSearch$.pipe(debounceTime(400)).subscribe((keyword) => {
+      dispatch(pathNavigatorSetKeyword({ id, keyword }));
+    });
```
Needs destruction

rart commented on ui/app/src/components/ToolsPanelEmbeddedAppViewButton/ToolsPanelEmbeddedAppViewButton.tsx:
```diff
@@ -31,7 +31,7 @@ const useStyles = makeStyles((theme) =>
     toolPanelBody: () => {
       const toolbarMixin: any = theme.mixins.toolbar;
-      const key1 = '@media (min-width:0) and (orientation: landscape)';
+      const key1 = '@media (min-width:0px) and (orientation: landscape)';
```
Was the `0` without px causing an issue?
A min-width of zero doesn't make a great deal of sense anyway. Wouldn't this work with just the orientation?

rart commented on ui/app/src/services/content.ts:
```diff
@@ -40,6 +40,7 @@ import { fetchContentTypes } from './contentTypes';
 import { Clipboard } from '../models/GlobalState';
 import { getPasteItemFromPath } from '../utils/path';
 import { StandardAction } from '../models/StandardAction';
+import { getStateMap } from '../utils/constants';
```
createItemStateMap

rart commented on ui/app/src/utils/constants.ts:
```diff
@@ -0,0 +1,61 @@
+const TRANSLATION_UP_TO_DATE_MASK = 0b0000000000000000000000000000000000000000000000000000001000000000;
+const TRANSLATION_PENDING_MASK = 0b0000000000000000000000000000000000000000000000000000010000000000;
+const TRANSLATION_IN_PROGRESS_MASK = 0b0000000000000000000000000000000000000000000000000000100000000000;
+const SUBMITTED_MASK = 0b0000000000000000000000000000000000000000000000000001000000000000;
```
- Name all as STATE_MASK_{state name} (e.g. STATE_MASK_NEW, etc)
- export them all
- I don't think we need all the leading zeros after the `0b`
- have a look around other utils to see if there are constants that should be moved here

aachavarria replied:
not sure about this `I don't think we need all the leading zeros after the 0b`

rart commented on ui/app/src/utils/constants.ts:
```diff
@@ -0,0 +1,61 @@
+    translationPending: isTranslationPending(status),
+    translationInProgress: isTranslationInProgress(status)
+  };
+};
```
Move to utils/content
Rename fns as is{state name}State
Rename value to stateBitMap
==> `createItemStateMap`

aachavarria committed:
ItemMenu.tsx: Fixing validation on noOptions using flatMap;
PathNavigator.tsx adding unsubscribe();
PathNavigatorList.tsx using item.path;
GetChildrenOptions.ts: adding new prop skipHomePathOverride;
SingleItemSelector.tsx: Changes to support new getChildrenAPI;
PR Feedback;

aachavarria committed:
removing taxonomy from itemActions Type;
adding console.error;

aachavarria committed:
Adding pagination to SingleItemSelector.tsx;

aachavarria committed:
PluginManagement Basic UI

rart commented on ui/app/src/models/Plugin.ts:
```diff
@@ -0,0 +1,32 @@
+/*
```
Add "Marketplace" to names, including file name

aachavarria replied:
will create another PR for this

rart commented on ui/app/src/utils/path.ts:
```diff
@@ -57,29 +56,6 @@ export function parseQueryString(): ParsedQuery {
   return parse(window.location.search);
 }
 
-// TODO: an initial path with trailing `/` breaks
-export function itemsFromPath(path: string, root: string, items: LookupTable<DetailedItem>): DetailedItem[] {
-  const rootWithIndex = withIndex(root);
```
Where did this go?

aachavarria replied:
This is not needed now, because we are using getIndividualPaths(), and only saving the path/url on the state

rart commented on ui/app/src/components/PluginManagement/PluginManagement.tsx:
```diff
@@ -0,0 +1,157 @@
+/*
```
Not sure what this is doing on this PR 🤔

aachavarria replied:
it should no be here

aachavarria committed:
Revert "PluginManagement Basic UI"

This reverts commit 381e5907

rart commented on ui/app/src/utils/constants.ts:
```diff
@@ -0,0 +1,29 @@
+export const STATE_MASK_TRANSLATION_UP_TO_DATE_MASK = 0b0000000000000000000000000000000000000000000000000000001000000000;
+export const STATE_MASK_TRANSLATION_PENDING_MASK = 0b0000000000000000000000000000000000000000000000000000010000000000;
+export const STATE_MASK_TRANSLATION_IN_PROGRESS_MASK = 0b0000000000000000000000000000000000000000000000000000100000000000;
+export const STATE_MASK_SUBMITTED_MASK = 0b0000000000000000000000000000000000000000000000000001000000000000;
```
```suggestion
export const STATE_MASK_NEW_MASK = 0b1;
export const STATE_MASK_MODIFIED_MASK = 0b10;
export const STATE_MASK_DELETED_MASK = 0b100;
export const STATE_MASK_LOCKED_MASK = 0b1000;
export const STATE_MASK_SYSTEM_PROCESSING_MASK = 0b10000;
export const STATE_MASK_IN_WORKFLOW_MASK = 0b100000;
export const STATE_MASK_SCHEDULED_MASK = 0b1000000;
export const STATE_MASK_STAGED_MASK = 0b10000000;
export const STATE_MASK_LIVE_MASK = 0b100000000;
export const STATE_MASK_TRANSLATION_UP_TO_DATE_MASK = 0b1000000000;
export const STATE_MASK_TRANSLATION_PENDING_MASK = 0b10000000000;
export const STATE_MASK_TRANSLATION_IN_PROGRESS_MASK = 0b100000000000;
export const STATE_MASK_SUBMITTED_MASK = 0b1000000000000;
```

rart committed:
Update ui/app/src/utils/constants.ts

Discard leading zeros
